src/lib/Err.hs

```diff
@@ -42,61 +54,142 @@ data ErrType = NoErr
              | ZipErr
              | EscapedNameErr
              | ModuleImportErr
+             | MonadFailErr
                deriving (Show, Eq)
 
-type Except = Either Err
+type SrcPosCtx  = Maybe SrcPos
+type SrcTextCtx = Maybe (Int, String) -- Int is the offset in the source file
+data ErrCtx = ErrCtx
+  { srcTextCtx :: SrcTextCtx
+  , srcPosCtx  :: SrcPosCtx
+  , messageCtx :: [String] }
+    deriving (Show, Eq)
+
+type Except = Either Errs
 
 type SrcPos = (Int, Int)
-type SrcCtx = Maybe SrcPos
 
-type MonadErr = MonadError Err
+class MonadFail m => Fallible m where
+  throwErrs :: Errs -> m a
+  addErrCtx :: ErrCtx -> m a -> m a
```

Comment:
What does adding context mean here? Error contexts don't seem to be forming a stack so it really seems like a local replacement of the context?

Reply:
The implementations are supposed to extend the context using the monoid instance of ErrCtx rather than replacing it. That might deserve a comment.

But your mention of a "stack" made me realize: with a suitably expanded `ErrCtx`, this `Fallible`/`CtxReader` are exactly what we need to accumulate a runtime stack in Simplify so that we can report pre-inlining stack traces at runtime!